### PR TITLE
Refactor date parsing in Refman and EndNote formats

### LINKED ISSUE #10: Parser should be more robust in injecting dates
In the particular case of RIS, "PY" is the Publication year which can also have the month and the day (e.g., YYYY/MM/DD or YYYYMMDD).
We should try our best effort to recognize the date and if all fail then put a null value.

### PR DESCRIPTION
**Pull request by kohrVid**
Closes #10.

This PR refactors the way in which dates are parsed in the Refman and EndNote plugins. In particular it:
* imports the correct date when it is split between the `DA` and `PY` tags in Refman files
* imports the correct date when both the `%8` and `%D` tags are populated in EndNote files
* parses the `%7` tag used for Edition dates when this is included in the EndNote files

## Questions

* ~For convenience, the order of preference for the date tags in the Refman format has been changed slightly so now the `DA` tag is checked before the `Y1`/`Y2`/`Y3` tags. Changing the order makes it possible to assign the `date_line` variable in a single one-liner without as opposed to three separate lines and removes the need for reassignment. Is there a particular reason as to why the `Y` tags were originally checked first or are the changes that I've made to that part of the code ok?~

### Refman

- [x] PY - YYYY/MM/D
- [x] DA - YYYY-MM-D
- [x] DA - YYYY MM D
- [x] Y1 - YYYY; Y2 - MM; Y3 - D
- [x] ---
- [x] DA - Dec 18; PY - 2012
- [x] DA - Dev 18 2012; PY - 2012
- [x]  DA - YYYY/MM; PY - YYYY
- [x] DA - 2020/03; PY - 2021
- [x] DA - SEP-OCT; PY - 2015
- [x] DA - WIN; PY - 1999
- [x] Y1 - YYYY/MM//
- [x] Y1 - YYYY/MM/DD/
- [x] DA - 2017/00
- [x] Y1 - YYYY; Y2 - MM	
- [x] Y1 - 2012/04//2012 Apr
- [x] Y1 - 2015/07/02/Jul2015 Supplement

### EndNote

- [x] %D YYYY
- [x] %D 2010; %7 2010/06/04; %8 Jun 1
- [x] %7 YYYY/MM/DD
- [x] %8 MM/D
- [x] --
- [x] %8 YYYY-MM-
- [x] %8 YYY-MM-DDThh:mm:ssZ
- [x] %D Fal; %8 Fal
- [x] %D YYYY; %8 YYYY-MM-
- [x] %D 2010; %8 Jun
- [x] %D Dec; %8 Dec-12-15
- [x] - [x] %D YYYY; %8 YYYY-MM
- [x] %8 YYYY--
- [x] %8 -1-1
- [x] %8 .1.1
- [x] %D 0001; %8 0001-01-01
- [x] %D September; %8 September
- [x] %D 25; %8 25-Apr
- [x] %D 2022; %8 September
- [x] %D YYYY; %8 YYYY.MM
- [x] %D YYYY; %8 YYYY-M-D
- [x] %D YYYY; %8 DD.MM.YYYY

### CODE REVIEW

File: lib/rayyan-formats-plugins/plugins/endnote.rb [hunk 1]
```diff
@@ -69,14 +70,22 @@ class EndNote < RayyanFormats::Base
...
           "\n"
         ].flatten.join if target
       end
 
       class << self
         def get_date_array(article)
-          date_line = article['8'] || article['D']
-          date_line.split(/[\/\s-]/)[0..2] if date_line
+          date_line = article['8'] && article['D'] ? "#{article['8']} #{article['D']}" : article['8'] || article['D']
+          date_array(date_line)
```
Comment (hammady) [inline lib/rayyan-formats-plugins/plugins/endnote.rb:81]: only `if date_line`

File: lib/rayyan-formats-plugins/plugins/endnote.rb [hunk 2]
```diff
@@ -102,6 +111,17 @@ def emit_line(key, value)
           "%#{key} #{value}\n" unless value.nil? || value.to_s.strip == ''
         end
 
+        private
+
+        def date_array(date_line)
```
Comment (hammady) [inline lib/rayyan-formats-plugins/plugins/endnote.rb:116]: Can you please explain what the logic here is? I am finding trouble reasoning it.
For example, if the length is less than or equal 4, why does this denote the year? It could be "June", "12/1", "12" or anything.
In the second case, you are replacing all special characters with `-`? Then why finally split on those 3 characters if you know the separator is space?
What if the `Date.parse` fails?

File: lib/rayyan-formats-plugins/plugins/endnote.rb [hunk 3]
```diff
@@ -41,6 +41,7 @@ class EndNote < RayyanFormats::Base
           target.abstracts = [article['X']].flatten.compact
           target.notes = try_join_arr article['1']
           target.doi = article['R']&.delete_prefix('doi:')
+          target.edition_date_array = get_edition_date_array article
```
Comment (hammady) [inline lib/rayyan-formats-plugins/plugins/endnote.rb:44]: This new field is not consumed in current clients of this gem. I suggest using this line as a fallback for `date_array` if it is blank.

File: lib/rayyan-formats-plugins/plugins/endnote.rb [hunk 4]
```diff
@@ -69,14 +70,22 @@ class EndNote < RayyanFormats::Base
...
+          target.edition_date_array ? emit_line("7", target.edition_date_array.join("-")) : nil,
           "\n"
         ].flatten.join if target
       end
 
       class << self
         def get_date_array(article)
-          date_line = article['8'] || article['D']
-          date_line.split(/[\/\s-]/)[0..2] if date_line
+          date_line = article['8'] && article['D'] ? "#{article['8']} #{article['D']}" : article['8'] || article['D']
```
Comment (hammady) [inline lib/rayyan-formats-plugins/plugins/endnote.rb:80]: `example2.enw` and `example3.enw` both have `%8` and `%D`, how would that work here? We cannot always combine them as they may be conflicting too.

File: lib/rayyan-formats-plugins/plugins/endnote.rb (hunk 2 above)
Comment (kohrVid) [reply lib/rayyan-formats-plugins/plugins/endnote.rb:116]: Those are all points that I wish I'd considered when I raised this PR. To be honest, I think I was having trouble figuring out what to return if `Date.parse` failed and so tried to write code that would prevent it reaching that point. `Date.parse` was considered as a straight-forward way to convert dates written as words (and possibly other unexpected formats) into something that could be split so I didn't think it mattered too much which character was used as a separator.

I wonder if a method like the following might work better?:

        def date_array(date_line)
          if date_line
            if !date_line.match(/[A-Za-z]/).nil?
              return Date.parse(date_line).strftime("%Y %-m %-d").split(/[\/\s-]/)[0..2]
            end

            Date.parse(date_line).strftime("%Y %-m %-d").split(/[\/\s-]/)[0..2] rescue date_line.split(/\D+/)
          end
        end

This could return unexpected results if the year is input as two digits or if the date is written in the European format but without the year ("DD/MM" instead of "MM/DD"). I'm not sure if we'd want to take that into account as well?

File: lib/rayyan-formats-plugins/plugins/endnote.rb (hunk 4 above)
Comment (kohrVid) [reply lib/rayyan-formats-plugins/plugins/endnote.rb:80]: When you say "conflicting", do you mean that there's a chance that the dates could be completely different in both (e.g., two different years for the same record)? My code assumes `%D` is always a year and so relied on `Date.parse()` to handle the instances where `%8` also contains a year. However, I'm not sure what we should do if the years are different.
This also doesn't take into account different formats for `%8` so perhaps I should parse that before combining the two values? Most likely the one-liner will have to go

Comment (hammady) [reply lib/rayyan-formats-plugins/plugins/endnote.rb]: I think the real user inputs I shared earlier might help you answer those questions.

**Comment by hammady** [comment]
The second link in the EndNote `description` field is invalid. Can you please replace it with a better one?

File: lib/rayyan-formats-plugins/plugins/endnote.rb (hunk 2 above)
Comment (hammady) [reply lib/rayyan-formats-plugins/plugins/endnote.rb]: The logic is becoming hard to follow. I suggest that you document all the cases you are covering in the PR description as checkboxes [ ] and [x]. Once you are done with implementing all, let me know and we can merge.

Comment (kohrVid) [reply lib/rayyan-formats-plugins/plugins/endnote.rb]: Have updated the PR description with a summary of the different date types that the plugins handle

Comment (hammady) [reply lib/rayyan-formats-plugins/plugins/endnote.rb]: Sounds good, what is this item `---`? If the code assumes no date in this case, then we should mark it as done with a comment. Let me know when you have fixed the unmarked ones.

File: spec/plugins/endnote_spec.rb [hunk 5]
```diff
@@ -43,14 +44,59 @@
...
           expect(target.abstracts).to eq(["abstract1"])
           expect(target.notes).to eq("notes1")
-        else
+        when 1
           expect(target.publication_types).to eq(["Another type"])
           expect(target.journal_title).to eq("journal2")
           expect(target.authors).to eq(["a1l, a1f", "a2l, a2f", "a3l, a3f", "a4l, a4f"])
           expect(target.url).to eq("url2")
           expect(target.doi).to eq("12.3456/science.124567")
+          expect(target.date_array).to eq(["2010", "6", "1"])
```
[the remark this answers is not in the document]
Comment (kohrVid) [reply spec/plugins/endnote_spec.rb:55]: I've added some comments to the file in the latest commit. Hopefully that's ok

File: spec/plugins/endnote_spec.rb [hunk 6]
```diff
@@ -50,50 +51,72 @@
...
+          # Date value(s): %D Fal; %8 Fal
           expect(target.date_array).to eq(nil)
         when 9
+          # Date value(s): %D 2022; %8 2021-09-
           expect(target.date_array).to eq(["2021", "9"])
         when 10
+          # Date value(s): %D 2010; %8 Jun
           expect(target.date_array).to eq(["2010", "6"])
         when 11
+          # Date value(s): %D Dec; %8 Dec-12-15
```
Comment (hammady) [inline spec/plugins/endnote_spec.rb:84]: Is this 12 Dec, 2015?

Comment (kohrVid) [reply spec/plugins/endnote_spec.rb:84]: > Is this 12 Dec, 2015?

I thought that particular string indicated a range of dates (as in "from the 12th December to the 15th December") but I suppose the "15" could be the year as well. I don't really know what we should expect in this case but if it's ambiguous, I wonder if it's safer to just treat it as invalid? Would users normally expect the last two digits in a date like this to refer to the year?

Comment (hammady) [reply spec/plugins/endnote_spec.rb:84]: @kohrVid

Comment (kohrVid) [reply spec/plugins/endnote_spec.rb:84]: I've updated the PR so now "Dec-12-15" is treated as "12th December 2015". It wasn't a particularly difficult change but I do worry that it could increase the number of incorrect years recorded. The `.date_array` methods depend on `Date.parse` which will assume the current year when it's missing in a string. I think it's harder to spot when this has happened with a two-digit year than a four-digit one. Is that an acceptable amount of risk in this case?

Comment (hammady) [reply spec/plugins/endnote_spec.rb:84]: Let's not take the risk.

**Review by hammady — CHANGES_REQUESTED** [review]

File: spec/plugins/endnote_spec.rb (hunk 6 above)
Comment (kohrVid) [reply spec/plugins/endnote_spec.rb:84]: Cool. I've just pushed a commit to revert those changes

**Review by hammady — APPROVED** [review]
